pkg/vulnsrc/openeuler/openeuler.go

```diff
@@ -0,0 +1,282 @@
+			if _, ok := uniqOSVers[pkg.OSVer]; !ok {
+				uniqOSVers[pkg.OSVer] = struct{}{}
+				if err := vs.dbc.PutDataSource(tx, pkg.OSVer, source); err != nil {
+					return xerrors.Errorf("failed to put data source: %w", err)
+				}
+			}
+
+			if err := vs.dbc.PutAdvisoryDetail(tx, cvrf.Tracking.ID, pkg.Name,
+				[]string{pkg.OSVer}, advisory); err != nil {
+				return xerrors.Errorf("unable to save %s CVRF: %w", pkg.OSVer, err)
+			}
+		}
+
+		var references []string
+		for _, ref := range cvrf.References {
+			references = append(references, ref.URL)
+		}
+
+		severity := types.SeverityUnknown
+		for _, cvuln := range cvrf.Vulnerabilities {
+			for _, threat := range cvuln.Threats {
+				sev := severityFromThreat(threat.Severity)
+				if severity < sev {
+					severity = sev
+				}
+			}
+		}
+
+		vuln := types.VulnerabilityDetail{
+			References:  references,
+			Title:       cvrf.Title,
+			Description: getDetail(cvrf.Notes),
+			Severity:    severity,
+		}
+		if err := vs.dbc.PutVulnerabilityDetail(tx, cvrf.Tracking.ID, source.ID, vuln); err != nil {
+			return xerrors.Errorf("failed to save openEuler CVRF vulnerability: %w", err)
+		}
+
+		// for optimization
+		if err := vs.dbc.PutVulnerabilityID(tx, cvrf.Tracking.ID); err != nil {
+			return xerrors.Errorf("failed to save the vulnerability ID: %w", err)
+		}
+	}
+	return nil
+}
+
+func getAffectedPackages(productTree ProductTree) []Package {
+	var pkgs []Package
+	var osArches = make(map[string][]string) // OS version => arches
+	for _, branch := range productTree.Branches {
+		// `src` pkgs are the really affected pkgs.
+		if branch.Type != "Package Arch" || branch.Name == "" {
+			continue
+		}
+		for _, production := range branch.Productions {
+			osVer := getOSVersion(production.CPE)
+			if osVer == "" {
+				log.Printf("Unable to parse OS version: %s", production.CPE)
+				continue
+			}
+
+			// Store possible architectures for OS version.
+			// We need this to find affected architectures for src pkg later.
+			if branch.Name != "src" {
+				if arches, ok := osArches[osVer]; ok {
+					osArches[osVer] = append(arches, branch.Name)
+				} else {
+					osArches[osVer] = []string{branch.Name}
+				}
+				continue
+			}
+
+			// e.g., `ignition-2.14.0-2` or `ignition-2.14.0-2.oe2203sp2.src.rpm`
+			pkgName, pkgVersion := parseProduction(production)
+			if pkgName == "" || pkgVersion == "" {
+				log.Printf("Unable to parse Production: %s", production)
+				continue
+			}
+			pkg := Package{
+				Name:         pkgName,
+				FixedVersion: pkgVersion,
+				OSVer:        osVer,
+			}
+			pkgs = append(pkgs, pkg)
+		}
+	}
+
+	// Fill affected architectures
+	for i, pkg := range pkgs {
+		arches := lo.Uniq(osArches[pkg.OSVer])
+		sort.Strings(arches)
+		pkgs[i].Arches = arches
+	}
+
+	return pkgs
+}
+
+func getOSVersion(cpe string) string {
+	// e.g. cpe:/a:openEuler:openEuler:22.03-LTS-SP3
+	parts := strings.Split(cpe, ":")
+	if len(parts) != 5 || parts[2] != "openEuler" {
+		return ""
+	}
```

Comment:
@wjunLu I found an incorrect CPE format - https://github.com/aquasecurity/vuln-list/blob/ecca778da16828205caa70ba6941582f70041104/openeuler/2022/openEuler-SA-2022-1773.json#L97

I checked other advisories. This format only appears in this file.
Do we need to parse this case or is it a typo by `openEular` and they should fix it on their end?

Reply:
Yes, I confirmed that some files exist such typo, and those errors occurred in the earlier advisories. 

I will try to ask the community to fix this, but I think those advisories that have been released may not be easy to modify.

Reply:
> I think those advisories that have been released may not be easy to modify.

I hope that it is possible to correct one advisory.

If no - we must provide for such cases

Reply:
> If no - we must provide for such cases
How to do this? list such cases in README or some other place?

Reply:
no, i meant we need to update our logic to add the ability to get the OS version for these cases.

Reply:
@DmitriyLewen Thanks! 
Please see
https://github.com/aquasecurity/trivy-db/blob/4fa6f19ef51011ed765f970c87b48611179e60c6/pkg/vulnsrc/openeuler/openeuler.go#L197

I added codes to use the wrong format getting the OS version, is that OK?